Index: linux-3.10/fs/ext4/ext4_jbd2.c
===================================================================
--- linux-3.10.orig/fs/ext4/ext4_jbd2.c	2016-04-26 05:53:15.554136137 +0200
+++ linux-3.10/fs/ext4/ext4_jbd2.c	2016-04-26 05:53:15.550802805 +0200
@@ -5,6 +5,7 @@
 #include "ext4_jbd2.h"
 
 #include <trace/events/ext4.h>
+#include <linux/cause_tags.h>
 
 /* Just increment the non-pointer handle value */
 static handle_t *ext4_get_nojournal(void)
@@ -198,6 +199,14 @@
 {
 	int err = 0;
 
+#ifndef DISABLE_CAUSES
+	// add this cause (meta write)
+	spin_lock(&bh->causes_lock);
+	cause_list_add(&bh->causes, current);
+	set_cause_list_type(bh->causes, SPLIT_CHECKPOINT);
+	spin_unlock(&bh->causes_lock);
+#endif
+
 	if (ext4_handle_valid(handle)) {
 		err = jbd2_journal_get_create_access(handle, bh);
 		if (err)
