Index: linux-3.9/fs/inode.c
===================================================================
--- linux-3.9.orig/fs/inode.c	2016-04-25 19:56:47.005402487 +0200
+++ linux-3.9/fs/inode.c	2016-04-25 19:56:47.002069155 +0200
@@ -219,6 +219,8 @@
 		return NULL;
 	}
 
+	inode->i_private1 = 0;
+
 	return inode;
 }
 
@@ -1511,7 +1513,9 @@
 		inode->i_ctime = *time;
 	if (flags & S_MTIME)
 		inode->i_mtime = *time;
+#ifndef SPLIT_NODEP
 	mark_inode_dirty_sync(inode);
+#endif
 	return 0;
 }
 
