Index: linux-3.10/include/linux/sched.h
===================================================================
--- linux-3.10.orig/include/linux/sched.h	2016-04-26 05:54:37.900800261 +0200
+++ linux-3.10/include/linux/sched.h	2016-04-26 05:54:37.897466929 +0200
@@ -1416,6 +1416,13 @@
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
