Index: linux-3.10/fs/xfs/xfs_vnodeops.c
===================================================================
--- linux-3.10.orig/fs/xfs/xfs_vnodeops.c	2016-04-26 05:54:16.357467594 +0200
+++ linux-3.10/fs/xfs/xfs_vnodeops.c	2016-04-26 05:54:16.350800927 +0200
@@ -596,6 +596,8 @@
 	 * the transaction cancel unlocking dp so don't do it explicitly in the
 	 * error path.
 	 */
+	//yangsuli
+	//here is where metadata actually got dirtied....
 	xfs_trans_ijoin(tp, dp, XFS_ILOCK_EXCL);
 	unlock_dp_on_error = false;
 
