Index: linux-3.11/fs/jbd2/checkpoint.c
===================================================================
--- linux-3.11.orig/fs/jbd2/checkpoint.c	2016-04-26 10:16:47.753706077 +0200
+++ linux-3.11/fs/jbd2/checkpoint.c	2016-04-26 10:16:47.750372743 +0200
@@ -322,6 +322,7 @@
 		(*batch_count)++;
 		if (*batch_count == JBD2_NR_BATCH) {
 			spin_unlock(&journal->j_list_lock);
+			printk("yangsuli: journal checkpoint process submit %d blocks\n", *batch_count);
 			__flush_batch(journal, batch_count);
 			ret = 1;
 		}
