Index: linux-3.7/include/linux/fs.h
===================================================================
--- linux-3.7.orig/include/linux/fs.h	2016-04-25 18:18:47.061065657 +0200
+++ linux-3.7/include/linux/fs.h	2016-04-25 18:18:47.054398991 +0200
@@ -606,6 +606,9 @@
 	atomic_t		i_readcount; /* struct files open RO */
 #endif
 	void			*i_private; /* fs or device private pointer */
+
+	// for scheduler to use as it pleases
+	long i_private1; // (maybe use as last write offset)
 };
 
 static inline int inode_unhashed(struct inode *inode)
