Index: linux-3.7/include/linux/sched.h
===================================================================
--- linux-3.7.orig/include/linux/sched.h	2016-04-25 18:19:25.027731152 +0200
+++ linux-3.7/include/linux/sched.h	2016-04-25 18:19:24.997731153 +0200
@@ -1548,6 +1548,13 @@
 #ifdef CONFIG_UPROBES
 	struct uprobe_task *utask;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
