Index: linux-3.19/include/linux/mm_types.h
===================================================================
--- linux-3.19.orig/include/linux/mm_types.h	2016-04-26 14:44:56.963235002 +0200
+++ linux-3.19/include/linux/mm_types.h	2016-04-26 14:44:56.956568335 +0200
@@ -197,7 +197,13 @@
 #ifdef LAST_CPUPID_NOT_IN_PAGE_FLAGS
 	int _last_cpupid;
 #endif
+
+    // cause list for btrfs
+#ifndef DISABLE_CAUSES
+        void* locked_causes;
+#endif
 }
+
 /*
  * The struct page can be forced to be double word aligned so that atomic ops
  * on double words work. The SLUB allocator can make use of such a feature.
