Index: linux-3.19/mm/page-writeback.c
===================================================================
--- linux-3.19.orig/mm/page-writeback.c	2016-04-26 14:45:04.226568149 +0200
+++ linux-3.19/mm/page-writeback.c	2016-04-26 14:45:04.223234815 +0200
@@ -2035,6 +2035,7 @@
 		ret = generic_writepages(mapping, wbc);
 	return ret;
 }
+EXPORT_SYMBOL(do_writepages);
 
 /**
  * write_one_page - write out a single page and optionally wait on I/O
