Index: linux-3.19/fs/xfs/xfs_aops.c
===================================================================
--- linux-3.19.orig/fs/xfs/xfs_aops.c	2016-04-26 14:45:17.453234476 +0200
+++ linux-3.19/fs/xfs/xfs_aops.c	2016-04-26 14:45:17.449901143 +0200
@@ -37,6 +37,8 @@
 #include <linux/pagevec.h>
 #include <linux/writeback.h>
 
+#include <linux/cause_tags.h>
+
 void
 xfs_count_page_state(
 	struct page		*page,
@@ -454,6 +456,7 @@
 
 static inline int xfs_bio_add_buffer(struct bio *bio, struct buffer_head *bh)
 {
+	move_causes_bh_to_bio(bh, bio);
 	return bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));
 }
 
