Index: linux-3.16/block/bio.c
===================================================================
--- linux-3.16.orig/block/bio.c	2016-04-26 13:40:24.366667651 +0200
+++ linux-3.16/block/bio.c	2016-04-26 13:40:24.360000984 +0200
@@ -28,6 +28,7 @@
 #include <linux/mempool.h>
 #include <linux/workqueue.h>
 #include <linux/cgroup.h>
+#include <linux/cause_tags.h>
 #include <scsi/sg.h>		/* for struct sg_iovec */
 
 #include <trace/events/block.h>
@@ -247,6 +248,8 @@
 	struct bio_set *bs = bio->bi_pool;
 	void *p;
 
+	del_cause_list_list(bio->cll);
+
 	__bio_free(bio);
 
 	if (bs) {
