Index: linux-3.16/mm/filemap.c
===================================================================
--- linux-3.16.orig/mm/filemap.c	2016-04-26 13:42:22.659997951 +0200
+++ linux-3.16/mm/filemap.c	2016-04-26 13:44:55.463327365 +0200
@@ -1100,6 +1100,10 @@
 		if (!page)
 			return NULL;
 
+#ifndef DISABLE_CAUSES
+		page->locked_causes = NULL;
+#endif
+
 		if (WARN_ON_ONCE(!(fgp_flags & FGP_LOCK)))
 			fgp_flags |= FGP_LOCK;
 
