Index: linux-3.12/mm/filemap.c
===================================================================
--- linux-3.12.orig/mm/filemap.c	2016-04-26 10:55:38.100312980 +0200
+++ linux-3.12/mm/filemap.c	2016-04-26 10:55:38.093646313 +0200
@@ -794,6 +794,10 @@
 		page = __page_cache_alloc(gfp_mask);
 		if (!page)
 			return NULL;
+
+#ifndef DISABLE_CAUSES
+		page->locked_causes = NULL;
+#endif
 		/*
 		 * We want a regular kernel memory (not highmem or DMA etc)
 		 * allocation for the radix tree nodes, but we need to honour
