Index: linux-3.14/kernel/hrtimer.c
===================================================================
--- linux-3.14.orig/kernel/hrtimer.c	2016-04-26 12:13:11.440193642 +0200
+++ linux-3.14/kernel/hrtimer.c	2016-04-26 12:13:11.433526976 +0200
@@ -1642,6 +1642,7 @@
 	destroy_hrtimer_on_stack(&t.timer);
 	return ret;
 }
+EXPORT_SYMBOL_GPL(hrtimer_nanosleep);
 
 SYSCALL_DEFINE2(nanosleep, struct timespec __user *, rqtp,
 		struct timespec __user *, rmtp)
