Index: linux-3.14/fs/btrfs/super.c
===================================================================
--- linux-3.14.orig/fs/btrfs/super.c	2016-04-26 12:13:36.083526343 +0200
+++ linux-3.14/fs/btrfs/super.c	2016-04-26 12:13:36.076859677 +0200
@@ -66,6 +66,24 @@
 static const struct super_operations btrfs_super_ops;
 static struct file_system_type btrfs_fs_type;
 
+struct block_device *btrfs_inode_to_bdev(struct inode* inode){
+	struct btrfs_inode *btrfs_inode = BTRFS_I(inode);
+	struct btrfs_device *btrfs_device;
+
+
+	if(!btrfs_inode->root)
+		return NULL;
+	if(!btrfs_inode->root->fs_info)
+		return NULL;
+	if(!btrfs_inode->root->fs_info->fs_devices)
+		return NULL;
+	if(btrfs_inode->root->fs_info->fs_devices->num_devices != 1)
+		return NULL;
+
+	btrfs_device = list_first_entry(&btrfs_inode->root->fs_info->fs_devices->devices, struct btrfs_device, dev_list);
+	return btrfs_device->bdev;
+}
+
 static const char *btrfs_decode_error(int errno)
 {
 	char *errstr = "unknown";
