Index: linux-3.10/fs/xfs/xfs_aops.c
===================================================================
--- linux-3.10.orig/fs/xfs/xfs_aops.c	2016-04-26 05:54:11.027467758 +0200
+++ linux-3.10/fs/xfs/xfs_aops.c	2016-04-26 05:54:11.024134425 +0200
@@ -37,6 +37,8 @@
 #include <linux/pagevec.h>
 #include <linux/writeback.h>
 
+#include <linux/cause_tags.h>
+
 void
 xfs_count_page_state(
 	struct page		*page,
@@ -453,6 +455,7 @@
 
 static inline int bio_add_buffer(struct bio *bio, struct buffer_head *bh)
 {
+	move_causes_bh_to_bio(bh, bio);
 	return bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));
 }
 
