Index: linux-3.11/fs/xfs/xfs_inode.c
===================================================================
--- linux-3.11.orig/fs/xfs/xfs_inode.c	2016-04-26 10:17:50.123704478 +0200
+++ linux-3.11/fs/xfs/xfs_inode.c	2016-04-26 10:17:50.113704478 +0200
@@ -1559,6 +1559,8 @@
 			 * Mark the inode dirty so it will be logged and
 			 * moved forward in the log as part of every commit.
 			 */
+			//yangsuli
+			//here dirties data...
 			xfs_trans_log_inode(tp, ip, XFS_ILOG_CORE);
 		}
 
@@ -1682,6 +1684,11 @@
 	agi->agi_unlinked[bucket_index] = cpu_to_be32(agino);
 	offset = offsetof(xfs_agi_t, agi_unlinked) +
 		(sizeof(xfs_agino_t) * bucket_index);
+	//yangsuli
+	//here we dirty
+	//but note that here the dirtier created/modified bp?
+	//is it true that who ever calls trans_log_buf dirtied the buf?
+	//possible...
 	xfs_trans_log_buf(tp, agibp, offset,
 			  (offset + sizeof(xfs_agino_t) - 1));
 	return 0;
