Index: linux-4.2/include/linux/mm_types.h
===================================================================
--- linux-4.2.orig/include/linux/mm_types.h	2016-04-26 17:00:51.923264640 +0200
+++ linux-4.2/include/linux/mm_types.h	2016-04-26 17:00:51.916597974 +0200
@@ -196,7 +196,13 @@
 #ifdef LAST_CPUPID_NOT_IN_PAGE_FLAGS
 	int _last_cpupid;
 #endif
+
+    // cause list for btrfs
+#ifndef DISABLE_CAUSES
+        void* locked_causes;
+#endif
 }
+
 /*
  * The struct page can be forced to be double word aligned so that atomic ops
  * on double words work. The SLUB allocator can make use of such a feature.
