Index: linux-4.3/fs/xfs/xfs_aops.c
===================================================================
--- linux-4.3.orig/fs/xfs/xfs_aops.c	2016-04-27 09:16:45.485036940 +0200
+++ linux-4.3/fs/xfs/xfs_aops.c	2016-04-27 09:16:45.478370273 +0200
@@ -36,6 +36,8 @@
 #include <linux/pagevec.h>
 #include <linux/writeback.h>
 
+#include <linux/cause_tags.h>
+
 void
 xfs_count_page_state(
 	struct page		*page,
@@ -433,6 +435,7 @@
 
 static inline int xfs_bio_add_buffer(struct bio *bio, struct buffer_head *bh)
 {
+	move_causes_bh_to_bio(bh, bio);
 	return bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));
 }
 
