Index: linux-3.8/include/linux/mm_types.h
===================================================================
--- linux-3.8.orig/include/linux/mm_types.h	2016-04-25 19:36:15.328773853 +0200
+++ linux-3.8/include/linux/mm_types.h	2016-04-25 19:36:15.322107186 +0200
@@ -176,7 +176,13 @@
 #ifdef CONFIG_NUMA_BALANCING
 	int _last_nid;
 #endif
+
+    // cause list for btrfs
+#ifndef DISABLE_CAUSES
+        void* locked_causes;
+#endif
 }
+
 /*
  * The struct page can be forced to be double word aligned so that atomic ops
  * on double words work. The SLUB allocator can make use of such a feature.
