Index: linux-3.8/fs/xfs/xfs_vnodeops.c
===================================================================
--- linux-3.8.orig/fs/xfs/xfs_vnodeops.c	2016-04-25 19:37:17.895438587 +0200
+++ linux-3.8/fs/xfs/xfs_vnodeops.c	2016-04-25 19:37:17.888771921 +0200
@@ -829,6 +829,8 @@
 	 * the transaction cancel unlocking dp so don't do it explicitly in the
 	 * error path.
 	 */
+	//yangsuli
+	//here is where metadata actually got dirtied....
 	xfs_trans_ijoin(tp, dp, XFS_ILOCK_EXCL);
 	unlock_dp_on_error = B_FALSE;
 
