Index: linux-3.4/fs/btrfs/extent_io.h
===================================================================
--- linux-3.4.orig/fs/btrfs/extent_io.h	2016-04-25 14:27:54.951377367 +0200
+++ linux-3.4/fs/btrfs/extent_io.h	2016-04-25 14:28:36.304709423 +0200
@@ -201,6 +201,7 @@
 int unlock_extent_cached(struct extent_io_tree *tree, u64 start, u64 end,
 			 struct extent_state **cached, gfp_t mask);
 int try_lock_extent(struct extent_io_tree *tree, u64 start, u64 end);
+void print_page_info(struct page *page);
 int extent_read_full_page(struct extent_io_tree *tree, struct page *page,
 			  get_extent_t *get_extent, int mirror_num);
 int __init extent_io_init(void);
