Index: linux-3.4/mm/filemap.c
===================================================================
--- linux-3.4.orig/mm/filemap.c	2016-04-25 14:46:42.168009226 +0200
+++ linux-3.4/mm/filemap.c	2016-04-25 14:46:42.161342560 +0200
@@ -775,6 +775,10 @@
 		page = __page_cache_alloc(gfp_mask);
 		if (!page)
 			return NULL;
+
+#ifndef DISABLE_CAUSES
+		page->locked_causes = NULL;
+#endif
 		/*
 		 * We want a regular kernel memory (not highmem or DMA etc)
 		 * allocation for the radix tree nodes, but we need to honour
