Index: linux-3.4/mm/page-writeback.c
===================================================================
--- linux-3.4.orig/mm/page-writeback.c	2016-04-25 14:47:20.118008055 +0200
+++ linux-3.4/mm/page-writeback.c	2016-04-25 14:47:20.111341388 +0200
@@ -1879,6 +1879,7 @@
 		ret = generic_writepages(mapping, wbc);
 	return ret;
 }
+EXPORT_SYMBOL(do_writepages);
 
 /**
  * write_one_page - write out a single page and optionally wait on I/O
