--- linux-3.2.51/include/linux/sched.h	2013-09-10 02:57:42.000000000 +0200
+++ ../split-release/release/linux-3.2.51/include/linux/sched.h	2015-09-01 05:42:28.000000000 +0200
@@ -1576,6 +1576,13 @@
 #ifdef CONFIG_HAVE_HW_BREAKPOINT
 	atomic_t ptrace_bp_refcnt;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
diff -rubBN linux-3.2.51/kernel/hrtimer.c ../split-release/release/linux-3.2.51/kernel/hrtimer.c
