--- linux-3.2.51/fs/xfs/xfs_inode.c	2013-09-10 02:57:42.000000000 +0200
+++ ../split-release/release/linux-3.2.51/fs/xfs/xfs_inode.c	2015-09-01 05:42:49.000000000 +0200
@@ -1372,6 +1372,8 @@
 		if (ip->i_size != new_size) {
 			ip->i_d.di_size = new_size;
 			ip->i_size = new_size;
+			//yangsuli
+			//here dirties data...
 			xfs_trans_log_inode(*tpp, ip, XFS_ILOG_CORE);
 		}
 	}
@@ -1478,6 +1480,11 @@
 	agi->agi_unlinked[bucket_index] = cpu_to_be32(agino);
 	offset = offsetof(xfs_agi_t, agi_unlinked) +
 		(sizeof(xfs_agino_t) * bucket_index);
+	//yangsuli
+	//here we dirty
+	//but note that here the dirtier created/modified bp?
+	//is it true that who ever calls trans_log_buf dirtied the buf?
+	//possible...
 	xfs_trans_log_buf(tp, agibp, offset,
 			  (offset + sizeof(xfs_agino_t) - 1));
 	return 0;
@@ -2593,6 +2600,7 @@
 	if (error)
 		goto cluster_corrupt_out;
 
+	//yangsuli: here dirty
 	if (flags & SYNC_WAIT)
 		error = xfs_bwrite(bp);
 	else
diff -rubBN linux-3.2.51/fs/xfs/xfs_vnodeops.c ../split-release/release/linux-3.2.51/fs/xfs/xfs_vnodeops.c
