Index: linux-3.18/mm/filemap.c
===================================================================
--- linux-3.18.orig/mm/filemap.c	2016-04-26 14:26:16.173263746 +0200
+++ linux-3.18/mm/filemap.c	2016-04-26 14:26:16.166597078 +0200
@@ -1115,6 +1115,10 @@
 		if (!page)
 			return NULL;
 
+#ifndef DISABLE_CAUSES
+		page->locked_causes = NULL;
+#endif
+
 		if (WARN_ON_ONCE(!(fgp_flags & FGP_LOCK)))
 			fgp_flags |= FGP_LOCK;
 
