Index: linux-3.18/fs/jbd2/journal.c
===================================================================
--- linux-3.18.orig/fs/jbd2/journal.c	2016-04-26 14:26:29.713263398 +0200
+++ linux-3.18/fs/jbd2/journal.c	2016-04-26 14:26:29.706596731 +0200
@@ -49,6 +49,7 @@
 
 #include <asm/uaccess.h>
 #include <asm/page.h>
+#include <linux/cause_tags.h>
 
 #ifdef CONFIG_JBD2_DEBUG
 ushort jbd2_journal_enable_debug __read_mostly;
@@ -2373,6 +2374,8 @@
 
 static void journal_free_journal_head(struct journal_head *jh)
 {
+	put_cause_list(jh->causes);
+
 #ifdef CONFIG_JBD2_DEBUG
 	atomic_dec(&nr_journal_heads);
 	memset(jh, JBD2_POISON_FREE, sizeof(*jh));
@@ -2443,6 +2446,7 @@
 			goto repeat;
 		}
 
+		new_jh->causes = NULL;
 		jh = new_jh;
 		new_jh = NULL;		/* We consumed it */
 		set_buffer_jbd(bh);
