Index: linux-4.5/include/linux/buffer_head.h
===================================================================
--- linux-4.5.orig/include/linux/buffer_head.h	2016-04-27 15:37:09.095771801 +0200
+++ linux-4.5/include/linux/buffer_head.h	2016-04-27 19:17:54.231820513 +0200
@@ -75,6 +75,9 @@
 	struct address_space *b_assoc_map;	/* mapping this buffer is
 						   associated with */
 	atomic_t b_count;		/* users using this buffer_head */
+
+	spinlock_t causes_lock;
+	struct cause_list *causes;
 };
 
 /*
