Index: linux-4.4/include/linux/init_task.h
===================================================================
--- linux-4.4.orig/include/linux/init_task.h	2016-04-27 09:37:18.671671980 +0200
+++ linux-4.4/include/linux/init_task.h	2016-04-27 09:37:18.668338647 +0200
@@ -189,6 +189,8 @@
  */
 #define INIT_TASK(tsk)	\
 {									\
+	.causes     = NULL,                     \
+	.account_id     = ATOMIC_INIT(0),	\
 	.state		= 0,						\
 	.stack		= &init_thread_info,				\
 	.usage		= ATOMIC_INIT(2),				\
