Index: linux-4.5/include/linux/journal-head.h
===================================================================
--- linux-4.5.orig/include/linux/journal-head.h	2016-04-27 15:37:09.742438450 +0200
+++ linux-4.5/include/linux/journal-head.h	2016-04-27 17:12:13.032878143 +0200
@@ -101,6 +101,8 @@
 
 	/* Trigger type for the committing transaction's frozen data */
 	struct jbd2_buffer_trigger_type *b_frozen_triggers;
+
+	struct cause_list *causes;
 };
 
 #endif		/* JOURNAL_HEAD_H_INCLUDED */
