Index: linux-4.1/fs/xfs/xfs_aops.c
===================================================================
--- linux-4.1.orig/fs/xfs/xfs_aops.c	2016-04-26 15:19:08.023182401 +0200
+++ linux-4.1/fs/xfs/xfs_aops.c	2016-04-26 15:19:08.016515734 +0200
@@ -36,6 +36,8 @@
 #include <linux/pagevec.h>
 #include <linux/writeback.h>
 
+#include <linux/cause_tags.h>
+
 void
 xfs_count_page_state(
 	struct page		*page,
@@ -437,6 +439,7 @@
 
 static inline int xfs_bio_add_buffer(struct bio *bio, struct buffer_head *bh)
 {
+	move_causes_bh_to_bio(bh, bio);
 	return bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));
 }
 
