Index: linux-3.5/include/linux/blk_types.h
===================================================================
--- linux-3.5.orig/include/linux/blk_types.h	2016-04-25 17:36:50.524476698 +0200
+++ linux-3.5/include/linux/blk_types.h	2016-04-25 17:36:50.521143365 +0200
@@ -82,6 +82,8 @@
 
 	bio_destructor_t	*bi_destructor;	/* destructor */
 
+	struct cause_list_list *cll;
+
 	/*
 	 * We can inline a number of vecs at the end of the bio, to avoid
 	 * double allocations for a small number of bio_vecs. This member
