Index: linux-3.12/mm/page-writeback.c
===================================================================
--- linux-3.12.orig/mm/page-writeback.c	2016-04-26 10:55:59.600312429 +0200
+++ linux-3.12/mm/page-writeback.c	2016-04-26 10:55:59.593645763 +0200
@@ -2071,6 +2071,7 @@
 		ret = generic_writepages(mapping, wbc);
 	return ret;
 }
+EXPORT_SYMBOL(do_writepages);
 
 /**
  * write_one_page - write out a single page and optionally wait on I/O
