Index: linux-3.12/include/linux/sched.h
===================================================================
--- linux-3.12.orig/include/linux/sched.h	2016-04-26 10:56:11.293645463 +0200
+++ linux-3.12/include/linux/sched.h	2016-04-26 10:56:11.286978796 +0200
@@ -1407,6 +1407,13 @@
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
