Index: linux-3.3/fs/inode.c
===================================================================
--- linux-3.3.orig/fs/inode.c	2016-04-25 11:41:37.876484633 +0200
+++ linux-3.3/fs/inode.c	2016-04-25 11:42:21.476483286 +0200
@@ -229,6 +229,8 @@
 		return NULL;
 	}
 
+	inode->i_private1 = 0;
+
 	return inode;
 }
 
@@ -1539,7 +1541,10 @@
 		return;
 
 	inode->i_atime = now;
+
+#ifndef SPLIT_NODEP
 	mark_inode_dirty_sync(inode);
+#endif
 	mnt_drop_write(mnt);
 }
 EXPORT_SYMBOL(touch_atime);
@@ -1590,7 +1595,10 @@
 		inode->i_ctime = now;
 	if (sync_it & S_MTIME)
 		inode->i_mtime = now;
+
+#ifndef SPLIT_NODEP
 	mark_inode_dirty_sync(inode);
+#endif
 	mnt_drop_write_file(file);
 }
 EXPORT_SYMBOL(file_update_time);
