Index: linux-3.3/include/linux/mm_types.h
===================================================================
--- linux-3.3.orig/include/linux/mm_types.h	2016-04-25 11:48:27.109805330 +0200
+++ linux-3.3/include/linux/mm_types.h	2016-04-25 11:48:27.106471996 +0200
@@ -149,7 +149,13 @@
 	 */
 	void *shadow;
 #endif
+
+    // cause list for btrfs
+#ifndef DISABLE_CAUSES
+        void* locked_causes;
+#endif
 }
+
 /*
  * The struct page can be forced to be double word aligned so that atomic ops
  * on double words work. The SLUB allocator can make use of such a feature.
