Index: linux-3.3/include/linux/sched.h
===================================================================
--- linux-3.3.orig/include/linux/sched.h	2016-04-25 11:48:46.456471400 +0200
+++ linux-3.3/include/linux/sched.h	2016-04-25 11:48:46.449804732 +0200
@@ -1590,6 +1590,13 @@
 #ifdef CONFIG_HAVE_HW_BREAKPOINT
 	atomic_t ptrace_bp_refcnt;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
