Index: linux-4.2/fs/ext4/page-io.c
===================================================================
--- linux-4.2.orig/fs/ext4/page-io.c	2016-04-26 17:00:56.119931199 +0200
+++ linux-4.2/fs/ext4/page-io.c	2016-04-26 17:00:56.116597865 +0200
@@ -23,6 +23,7 @@
 #include <linux/kernel.h>
 #include <linux/slab.h>
 #include <linux/mm.h>
+#include <linux/cause_tags.h>
 
 #include "ext4_jbd2.h"
 #include "xattr.h"
@@ -409,6 +410,10 @@
 	ret = bio_add_page(io->io_bio, page, bh->b_size, bh_offset(bh));
 	if (ret != bh->b_size)
 		goto submit_and_retry;
+
+	// I think this path is only for writes
+	move_causes_bh_to_bio(bh, io->io_bio);
+
 	io->io_next_block++;
 	return 0;
 }
