Index: linux-4.2/include/linux/sched.h
===================================================================
--- linux-4.2.orig/include/linux/sched.h	2016-04-26 17:01:32.119930276 +0200
+++ linux-4.2/include/linux/sched.h	2016-04-26 18:01:11.409838482 +0200
@@ -1776,6 +1776,14 @@
 	unsigned long	task_state_change;
 #endif
 	int pagefault_disabled;
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
+
 /* CPU-specific state of this task */
 	struct thread_struct thread;
 /*
