Index: linux-4.0/include/linux/fs.h
===================================================================
--- linux-4.0.orig/include/linux/fs.h	2016-04-26 15:07:16.736533976 +0200
+++ linux-4.0/include/linux/fs.h	2016-04-26 15:07:16.733200642 +0200
@@ -639,6 +639,9 @@
 #endif
 
 	void			*i_private; /* fs or device private pointer */
+
+	// for scheduler to use as it pleases
+	long i_private1; // (maybe use as last write offset)
 };
 
 static inline int inode_unhashed(struct inode *inode)
