Index: linux-4.0/fs/xfs/xfs_aops.c
===================================================================
--- linux-4.0.orig/fs/xfs/xfs_aops.c	2016-04-26 15:07:54.529866340 +0200
+++ linux-4.0/fs/xfs/xfs_aops.c	2016-04-26 15:07:54.526533006 +0200
@@ -37,6 +37,8 @@
 #include <linux/pagevec.h>
 #include <linux/writeback.h>
 
+#include <linux/cause_tags.h>
+
 void
 xfs_count_page_state(
 	struct page		*page,
@@ -438,6 +440,7 @@
 
 static inline int xfs_bio_add_buffer(struct bio *bio, struct buffer_head *bh)
 {
+	move_causes_bh_to_bio(bh, bio);
 	return bio_add_page(bio, bh->b_page, bh->b_size, bh_offset(bh));
 }
 
