Index: linux-4.0/fs/xfs/xfs_inode.c
===================================================================
--- linux-4.0.orig/fs/xfs/xfs_inode.c	2016-04-26 15:02:14.996541714 +0200
+++ linux-4.0/fs/xfs/xfs_inode.c	2016-04-26 15:05:04.556537365 +0200
@@ -1561,6 +1561,8 @@
 			 * Mark the inode dirty so it will be logged and
 			 * moved forward in the log as part of every commit.
 			 */
+			//yangsuli
+			//here dirties data...
 			xfs_trans_log_inode(tp, ip, XFS_ILOG_CORE);
 		}
 
@@ -1996,6 +1998,11 @@
 	offset = offsetof(xfs_agi_t, agi_unlinked) +
 		(sizeof(xfs_agino_t) * bucket_index);
 	xfs_trans_buf_set_type(tp, agibp, XFS_BLFT_AGI_BUF);
+	//yangsuli
+	//here we dirty
+	//but note that here the dirtier created/modified bp?
+	//is it true that who ever calls trans_log_buf dirtied the buf?
+	//possible...
 	xfs_trans_log_buf(tp, agibp, offset,
 			  (offset + sizeof(xfs_agino_t) - 1));
 	return 0;
@@ -2088,6 +2095,11 @@
 		offset = offsetof(xfs_agi_t, agi_unlinked) +
 			(sizeof(xfs_agino_t) * bucket_index);
 		xfs_trans_buf_set_type(tp, agibp, XFS_BLFT_AGI_BUF);
+		//yangsuli
+		//here we dirty
+		//but note that here the dirtier created/modified bp?
+		//is it true that who ever calls trans_log_buf dirtied the buf?
+		//possible...
 		xfs_trans_log_buf(tp, agibp, offset,
 				  (offset + sizeof(xfs_agino_t) - 1));
 	} else {
