Index: linux-3.13/include/linux/sched.h
===================================================================
--- linux-3.13.orig/include/linux/sched.h	2016-04-26 11:05:42.576964145 +0200
+++ linux-3.13/include/linux/sched.h	2016-04-26 11:05:42.570297479 +0200
@@ -1454,6 +1454,13 @@
 	unsigned int	sequential_io;
 	unsigned int	sequential_io_avg;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
