Index: linux-3.7/include/linux/blk_types.h
===================================================================
--- linux-3.7.orig/include/linux/blk_types.h	2016-04-25 18:18:10.834400110 +0200
+++ linux-3.7/include/linux/blk_types.h	2016-04-25 18:18:10.831066776 +0200
@@ -86,6 +86,8 @@
 
 	struct bio_set		*bi_pool;
 
+	struct cause_list_list *cll;
+
 	/*
 	 * We can inline a number of vecs at the end of the bio, to avoid
 	 * double allocations for a small number of bio_vecs. This member
