Index: linux-3.9/include/linux/sched.h
===================================================================
--- linux-3.9.orig/include/linux/sched.h	2016-04-25 19:57:18.815401505 +0200
+++ linux-3.9/include/linux/sched.h	2016-04-25 19:57:18.798734839 +0200
@@ -1577,6 +1577,13 @@
 #ifdef CONFIG_UPROBES
 	struct uprobe_task *utask;
 #endif
+
+	// sometimes a task acts as a proxy for other threads (e.g.,
+	// a flush thread acting on behalf of other tasks that did
+	// delayed allocation).
+	struct cause_list* causes;
+
+	atomic_t account_id;
 };
 
 /* Future-safe accessor for struct task_struct's cpus_allowed. */
